Sources/Main.cpp

```diff
@@ -1268,5 +1280,126 @@ void __fastcall TX584Form::AboutItemClick(TObject *Sender)
 {
     AboutForm->ShowModal();
 }
+
+//---------------------------------------------------------------------------
+//                           *** БУФЕР ОБМЕНА ***
+//---------------------------------------------------------------------------
+
+int TX584Form::GetClipboardBufferSize()
+{
+    int result = 0;
+
+    result += 2; // количество элементов в буфере
+    for (int i = 0; i < ClipboardSize; i++) {
+		result += 2 + 1 + CMClipboard[i].Length() * 2;
+    }
+
+    return result;
+}
+//---------------------------------------------------------------------------
+
+void TX584Form::PutIntoClipboard()
+{
+    HGLOBAL hBuffer = NULL;
+    unsigned char *Buffer, *p;
+    int BufferSize;
+
+    Clipboard()->Open();
+    Clipboard()->Clear();
+    BufferSize = GetClipboardBufferSize();
+
+    hBuffer = GlobalAlloc(GMEM_MOVEABLE, BufferSize);
+    Buffer = (unsigned char *)GlobalLock(hBuffer);
+    p = Buffer;
+
+    // записать количество элементов в буфере обмена
+    *p++ = (ClipboardSize & 0xFF);
+    *p++ = ((ClipboardSize >> 8) & 0xFF);
+    // записать каждый элемент
+    for (int i = 0; i < ClipboardSize; i++) {
+        // записать микроинструкцию
+        *p++ = (MIClipboard[i] & 0xFF);
+        *p++ = ((MIClipboard[i] >> 8) & 0xFF);
+        // записать длину комментария
+        *p++ = (CMClipboard[i].Length() & 0xFF);
+        // записать комментарий
+		for (int j = 1; j <= CMClipboard[i].Length(); j++) {
+			WideChar Ch = CMClipboard[i][j];
+			*p++ = ((int)Ch)&0xFF;
+			*p++ = (((int)Ch) >> 8) & 0xFF;
+        }
+    }
+    GlobalUnlock(hBuffer);
+    Clipboard()->SetAsHandle(ClipboardFormat, (unsigned)hBuffer);
```

Comment:
Oops, this will crash in 64-bit builds. In a 64-bit compiler the size of `hBuffer` (i.e. pointer) is 64 bits whereas `unsigned` is still 32 bits. This means C-style cast will truncate `hBuffer` to 32 bits and the application will crash, probably without a single warning from the compiler.

That's why C-style casts are evil. One should never use them in modern C++ code. If absolutely necessary, `static_cast` or `reinterpret_cast` should be used instead. Here we should cast `hBuffer` to `THandle` which is 32 bit in 32-bit builds and 64 bit in 64-bit builds.

```cpp
Clipboard()->SetAsHandle(ClipboardFormat, reinterpret_cast<THandle>(hBuffer));
```